substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -672,50 +673,44 @@ impl<T: Config> Pallet<T> {
 		let mut source_delegation =
 			Delegators::<T>::get(&source_delegator).defensive_ok_or(Error::<T>::BadState)?;
 
-		// some checks that must have already been checked before.
+		// ensure source has enough funds to migrate.
 		ensure!(source_delegation.amount >= amount, Error::<T>::NotEnoughFunds);
 		debug_assert!(
 			!Self::is_delegator(&destination_delegator) && !Self::is_agent(&destination_delegator)
 		);
 
 		let agent = source_delegation.agent.clone();
-		// update delegations
-		Delegation::<T>::new(&agent, amount).update_or_kill(&destination_delegator);
+		// create a new delegation for destination delegator.
+		let _ = Delegation::<T>::new(&agent, amount).update(&destination_delegator);
+		// Provide for the delegator account. This ensures that amount including existential deposit
+		// can be held.
+		frame_system::Pallet::<T>::inc_providers(&destination_delegator);
 
 		source_delegation.amount = source_delegation
 			.amount
 			.checked_sub(&amount)
 			.defensive_ok_or(Error::<T>::BadState)?;
 
-		source_delegation.update_or_kill(&source_delegator);
+		// ensure amount is greater than ED otherwise transfer would fail.
+		ensure!(amount >= T::Currency::minimum_balance(), Error::<T>::NotEnoughFunds);
```

Comment:
I suggest you to make the error description of `NotEnoughFunds` and other ones that you foresee would be hit by users and wallet devs to be very descriptive.

Reply:
This error specifically should already be checked by nomination pool and only preventive here. But point taken, I will review the error descriptions again including the pool pallet.